internal/terraform/node_action_trigger_abstract.go

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package terraform
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/configs"
+	"github.com/hashicorp/terraform/internal/dag"
+	"github.com/hashicorp/terraform/internal/lang/langrefs"
+)
+
+type RelativeActionTiming = string
+
+const (
+	RelativeActionTimingBefore = "before"
+	RelativeActionTimingAfter  = "after"
+)
+
+// ConcreteActionTriggerNodeFunc is a callback type used to convert an
+// abstract action trigger to a concrete one of some type.
+type ConcreteActionTriggerNodeFunc func(*nodeAbstractActionTriggerExpand, RelativeActionTiming) dag.Vertex
+
+type nodeAbstractActionTriggerExpand struct {
+	Addr             addrs.ConfigAction
+	resolvedProvider addrs.AbsProviderConfig
+	Config           *configs.Action
+
+	lifecycleActionTrigger *lifecycleActionTrigger
+}
+
+type lifecycleActionTrigger struct {
+	resourceAddress         addrs.ConfigResource
+	events                  []configs.ActionTriggerEvent
+	actionTriggerBlockIndex int
+	actionListIndex         int
+	invokingSubject         *hcl.Range
+	actionExpr              hcl.Expression
+	conditionExpr           hcl.Expression
+}
+
+func (at *lifecycleActionTrigger) Name() string {
+	return fmt.Sprintf("%s.lifecycle.action_trigger[%d].actions[%d]", at.resourceAddress.String(), at.actionTriggerBlockIndex, at.actionListIndex)
+}
+
+var (
+	_ GraphNodeReferencer = (*nodeAbstractActionTriggerExpand)(nil)
```

Comment:
Can you add the other ones in here? Like GraphNodeModulePath and GraphNodeProviderConsumer? Whatever they are.